### PR TITLE
Bugfix: Fixed batch splitting.

### PR DESCRIPTION
**Pull request by edgar-tr**
The inner for loop (_subbatch) had no effect -- it always used the same subbatch. This can be verified by printing sdf_data or loss in every iteration of _subbatch. The outer for loop generates a subbatch (sdf_data) and this subbatch is then re-used _batch_split_ many times in the inner for loop.

I'm new to Pytorch, so maybe I'm using it wrong.

### CODE REVIEW

**Comment by edgar-tr** [comment]
Maybe to clarify: This is a bugfix, not a feature. In its current version, batch splitting is broken and should not be used.

Let's say we split the batch (batch_size = 64) into four subbatch of sizes 16 each (batch_split = 4). The goal is to accumulate the gradients of the four _different_ subbatches and then do an optimization step. What the current implementation does in this scenario is instead the following: we get a (sub-)batch of size 16, run this exact subbatch four times, accumulate four times the _same_ gradients, and then do an update step. No accumulation of _different_ gradients occurs, since all four subbatches are the _same_. 

In effect, current batch splitting increases the learning rate by a factor of four, reduces the batch size from 64 to 16, and increases the training time by a factor of four. Especially the last point is a major issue.

**Comment by HM102** [comment]
Did you implement the gradient accumulation?

**Comment by edgar-tr** [comment]
Batch splitting is a feature that DeepSDF is supposed to have. There is a bug with which gradients are accumulated, not with the accumulation itself. Accumulating gradients is straightforward in PyTorch from what I understand: as long as the gradient buffers are not explicitly set to 0 (optimizer_all.zero_grad()), loss.backward() calls will keep on adding/accumulating gradients in them. optimizer_all.step() will then use these gradient buffers to update the weights. Is that what you meant?

**Comment by HM102** [comment]
yes, that's what I meant. I was wondering if you already did the batch split  and saw a difference in the loss.

**Comment by edgar-tr** [comment]
No, I haven't trained it yet, I just started playing around with the code

**Comment by HM102** [comment]
> No, I haven't trained it yet, I just started playing around with the code



> Batch splitting is a feature that DeepSDF is supposed to have. There is a bug with which gradients are accumulated, not with the accumulation itself. Accumulating gradients is straightforward in PyTorch from what I understand: as long as the gradient buffers are not explicitly set to 0 (optimizer_all.zero_grad()), loss.backward() calls will keep on adding/accumulating gradients in them. optimizer_all.step() will then use these gradient buffers to update the weights. Is that what you meant?

Should the` loss` be divided by  `batch_split`, before doing `loss.backward()`?

**Comment by edgar-tr** [comment]
I fixed it now.

**Comment by tschmidt23** [comment]
Hey @edgar-tr,

Thanks for catching this! I had a look at this bug and decided to take the opportunity to move the storage of latent vectors from a list of torch.Tensor objects to a single torch.nn.Embedding object, which is something I've been meaning to do anyways. This change made other aspects of the code, including batch splitting, a bit simpler, so you will see that in the latest commit (d82529227732892a5ca16631c360c384a6e17372), the batch splitting has been fixed in a slightly different way (pulling a full batch from the data loader and then splitting it with torch.chunk). I hope this works for your use cases as well and please let me know of and concerns or comments!

**Comment by edgar-tr** [comment]
Cool, I haven't tested it, but that looks really good. Thank you!
